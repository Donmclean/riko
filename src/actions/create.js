import { regexReplaceCustomBoilerplateString, genericLog } from '../utils/functions';
import { cwd } from '../utils/variables';
import fs from 'fs-extra';
import regexReplace from 'regex-replace';

export default async (finalPath, actionType, fileType, fileName) => {
    const fullBinFilePath = `${finalPath}/${fileType}`;

    const filePathStats = await fs.stat(fullBinFilePath);

    //handle Directories
    if(filePathStats.isDirectory()) {
        try {
            await fs.copy(fullBinFilePath, `${cwd}/${fileName}`);
        } catch (err) {
            console.error(`ERROR > error creating directory (${cwd}/${fileName})`, err);
            return 1;
        }

        try {
            await regexReplace('<:rikofilename:>', fileName, `${cwd}/${fileName}`);
        } catch (err) {
            console.error(`ERROR > error replacing custom filenames in (${cwd}/${fileName})`, err);
            return 1;
        }

        genericLog(`${fileName} created successfully`);

    } else { //handle files
        let customizedFile;
        try {
            customizedFile = await fs.readFile(fullBinFilePath, 'utf8');
        } catch (err) {
            console.error(`ERROR > error reading file (${cwd}/${fileName})`, err);
            return 1;
        }

        const editedContent = regexReplaceCustomBoilerplateString(customizedFile, fileName);

        try {
            await fs.writeFile(`${cwd}/${fileName}`, editedContent);
        } catch (err) {
            console.error(`ERROR > error writing to file (${cwd}/${fileName})`, err);
            return 1;
        }

        genericLog(`${fileName} created successfully`);
    }
};