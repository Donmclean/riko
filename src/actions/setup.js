import { genericLog, folderAlreadyPresent, sortObjByOwnKeys } from '../utils/functions';
import { cwd, baseDir, packageJson } from '../utils/variables';
import qfs from 'q-io/fs';
import spawn from 'cross-spawn';
import gulpLoadPlugins from 'gulp-load-plugins';

const $ = gulpLoadPlugins();
const spawnSync = spawn.sync;

export default (actionType, projectType, projectName) => {
    const logSuccess = () => genericLog(`${$.util.colors.blue(projectName)} was setup ${$.util.colors.green('successfully')}`);

    return qfs.list(cwd)
        .then(files => {
            if(folderAlreadyPresent(files, projectName)) {
                genericLog(`${$.util.colors.blue(projectName)} folder must not exist during setup. ${$.util.colors.red('terminating...')}`);
                throw new Error(`${projectName} folder must not exist during setup.`);
            } else {
                switch (projectType) {
                    case 'electron': {
                        qfs.copyTree(`${baseDir}/bin/_${actionType}/react`, `${cwd}/${projectName}`)
                            .then(() => qfs.copyTree(`${baseDir}/bin/_${actionType}/${projectType}`, `${cwd}/${projectName}/src`))
                            .then(() => qfs.read(`${cwd}/${projectName}/package.json`))
                            .then((customPackageJson) => {
                                //adds electron as devDependencies to package.json
                                const packageJsonObj = JSON.parse(customPackageJson);
                                packageJsonObj.devDependencies = Object.assign(
                                    {},
                                    packageJsonObj.devDependencies,
                                    {electron: packageJson.dependencies.electron}
                                );
                                packageJsonObj.devDependencies = sortObjByOwnKeys(packageJsonObj.devDependencies);
                                return packageJsonObj;
                            })
                            .then((packageJsonObj) => qfs.write(`${cwd}/${projectName}/package.json`, JSON.stringify(packageJsonObj, null, "\t")))
                            .then(() => logSuccess())
                            .catch((err) => genericLog(err, 'red'));
                        break;
                    }
                    case 'node-server':
                    case 'react': {
                        qfs.copyTree(`${baseDir}/bin/_${actionType}/${projectType}`, `${cwd}/${projectName}`)
                            .then(() => logSuccess())
                            .catch((err) => genericLog(err, 'red'));
                        break;
                    }
                    case 'react-native': {
                        //run react native shell script
                        spawnSync('sh', [`${baseDir}/bin/_${actionType}/${projectType}/react-native-install.sh`, projectName], {stdio: 'inherit'});

                        //copy rikoconfig.js file
                        qfs.makeTree(`${cwd}/${projectName}/src`)
                            .then(() => qfs.copy(`${baseDir}/bin/_${actionType}/${projectType}/rikoconfig.js`, `${cwd}/${projectName}/src/rikoconfig.js`))
                            .then(() => genericLog(`${$.util.colors.blue(`${projectName}`)} folder created ${$.util.colors.green('successfully')}`))
                            .catch((err) => genericLog(err, 'red'));
                        break;
                    }
                    default: {
                        break;
                    }
                }
            }
        })
        .catch((err) => genericLog(err, 'red'));
};