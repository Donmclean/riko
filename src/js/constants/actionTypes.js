export const SHOULD_INCREMENT = 'SHOULD_INCREMENT';
export const SHOULD_DECREMENT = 'SHOULD_DECREMENT';
export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';

export const UPDATE_POSTS = 'UPDATE_POSTS';
export const UPDATE_POST_NUMBER = 'UPDATE_POST_NUMBER';