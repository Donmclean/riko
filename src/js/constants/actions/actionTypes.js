export const INCREMENT = 'INCREMENT';
export const DECREMENT = 'DECREMENT';

export const GET_POSTS = 'GET_POSTS';
export const UPDATE_INPUT_VALUE = 'UPDATE_INPUT_VALUE';