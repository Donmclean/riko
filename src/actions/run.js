import { processExitHandler, assignEnvironmentVariablesBasedOnRunCommand, doRunCommandValidations, logElectronRunServerError,
    removeDir, genericLog, getStats, onDevBuildActions, checkForNewPackageVersion } from '../utils/functions';
import { cwd, baseDir } from '../utils/variables';
import { getElectronPackagerOptions } from '../utils/webpackConfigUtils';
import qfs from 'q-io/fs';
import os from 'os';
import path from 'path';
import { find } from 'lodash';
import { runCommands } from '../constants/index';
import spawn from 'cross-spawn';
const spawnSync = spawn.sync;

export default (runCommand) => {
    //add exit handler
    processExitHandler();

    checkForNewPackageVersion();

    //assign environment variables
    assignEnvironmentVariablesBasedOnRunCommand(runCommands, runCommand);

    const requiresWebpack = (JSON.parse(process.env.isReact) || JSON.parse(process.env.isElectron));

    doRunCommandValidations();

    const modulesToImport = [ import('../utils/coreRikoConfig').then((m) => m.default) ];

    if(requiresWebpack) {
        modulesToImport.push(import('../webpack.config.babel').then((m) => m.default));
    }

    Promise.all(modulesToImport).then((modules) => {
        const [ customConfig, config = {} ] = modules;

        //TODO: validate customConfig Here
        //TODO: extract these dependencies

        switch (runCommand) {
            case 'electron-server': {
                qfs.list(customConfig.output.path).then((files) => {
                    switch (os.platform()) {
                        case 'darwin': {
                            //handles open electron app on MAC
                            const macFile = find(files, (file) => file.match(/\b(darwin)\b/i));

                            if(macFile) {
                                spawn('open', [`-a`, `${cwd}/${path.basename(customConfig.output.path)}/${macFile}/${customConfig.electronPackagerOptions.name}.app`], {stdio: 'inherit'});
                            } else {
                                logElectronRunServerError();
                            }

                            break;
                        }
                        default: {
                            break;
                        }
                    }
                })
                    .catch((err) => {
                        logElectronRunServerError();
                        console.error('ERROR >', err);
                    });
                break;
            }
            case 'react-prod': {
                config.entry['index'].unshift('babel-polyfill');

                //TODO: validate command below
                return spawn(`${customConfig.baseDir}/node_modules/.bin/webpack`, [`--config`, `${path.resolve(__dirname, '../webpack.config.babel.js')}`], {stdio: 'inherit'});
            }
            case 'electron-prod': {
                genericLog('Compiling electron app..');

                if(JSON.parse(process.env.isElectron)) {
                    // const electronPackager = require('electron-packager'); //TODO: use dynamic import here
                    return import('electron-packager')
                        .then((electronPackager) => {
                            const spawned = spawn(`${customConfig.baseDir}/node_modules/.bin/webpack`, [`--config`, `${path.resolve(__dirname, '../webpack.config.babel.js')}`], {stdio: 'inherit'});

                            spawned.on('close', () => {
                                //Compile The Electron Application
                                const electronPackagerOptions = getElectronPackagerOptions();
                                electronPackager(electronPackagerOptions, (err) => {
                                    if(err) {
                                        console.error('ERROR > in electron build', err);
                                        throw err;
                                    }

                                    removeDir(customConfig.tempDir).then(() => {
                                        genericLog(`${electronPackagerOptions.name} built successfully!`);
                                    });
                                });
                            });

                            return spawn;
                        });
                }
                break;
            }
            case 'react-native-ios': {
                return spawn('react-native', ['run-ios'], {stdio: 'inherit'});
            }
            case 'react-native-launch-android': {
                genericLog('launching android emulator...', 'blue');
                return spawn(`sh`, [`${cwd}/launch-android-Emulator.sh`, '&'], {stdio: 'ignore'});
            }
            case 'react-native-android': {
                return spawnSync('react-native', ['run-android'], {stdio: 'inherit'});
            }
            case 'electron-dev':
            case 'react-dev': {
                const stats = getStats('development');
                const webpack = require('webpack');
                const webpackDevServer = require('webpack-dev-server');

                //*******************************************************************
                //HOT RELOADING WITH WEBPACK DEV SERVER
                //*******************************************************************

                config.entry['index'].unshift('webpack/hot/dev-server');
                config.entry['index'].unshift(`webpack-dev-server/client?http://localhost:${customConfig.SERVER_PORT}`);
                config.entry['index'].unshift('react-hot-loader/patch');
                config.entry['index'].unshift('babel-polyfill');

                const { overlay } = customConfig.hotReloadingOptions;

                new webpackDevServer(webpack(config), {
                    contentBase: config.output.path,
                    publicPath: config.output.publicPath,
                    historyApiFallback: true,
                    hot: true,
                    overlay,
                    inline: true,
                    headers: { 'Access-Control-Allow-Origin': '*' },
                    stats
                }).listen(customConfig.SERVER_PORT, 'localhost', (err) => {
                    if (err) { console.error(err); }
                    genericLog(`Listening at localhost: ${customConfig.SERVER_PORT}`);
                    onDevBuildActions(customConfig);
                });

                break;
            }
            case 'node-server-dev': {
                spawn(`${baseDir}/node_modules/.bin/nodemon`, ['--config', `${customConfig.nodemonJson}`, `${customConfig.entryFile}`], {stdio: 'inherit'});
                break;
            }
            case 'node-server-prod': {
                spawn('node', [`${customConfig.entryFile}`], {stdio: 'inherit'});
                break;
            }
            case 'react-server':
            case 'react-prod-server': {
                //TODO: use dynamic import
                const browserSync   = require('browser-sync');
                const morgan        = require('morgan');
                const fallback      = require('express-history-api-fallback');
                const express       = require('express');
                const app           = require('express')();

                app.use(morgan('dev'));

                const root = config.output.path;

                app.use(express.static(root));
                app.use(fallback('index.html', { root }));

                const portToServe = process.argv[4] || customConfig.SERVER_PORT;

                app.listen(portToServe);

                app.use((err, req, res, next) => {
                    genericLog(`ERROR --> : ${err.stack}`);
                    next(err);
                });

                const isProdServer = (runCommand === 'react-prod-server');

                if(isProdServer) {
                    genericLog('Listening on port: ' + portToServe, 'yellow');
                } else {
                    genericLog('Launching Browser Sync proxy of port: ' + portToServe, 'yellow');

                    browserSync.init({
                        proxy: `localhost:${portToServe}`
                    });
                }

                break;
            }
            default: {
                break;
            }
        }
    });
};