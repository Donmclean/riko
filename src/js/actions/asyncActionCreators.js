"use strict";
import axios from 'axios';
import * as types from '../../js/constants/actions/actionTypes';

export const updateInputValue = (postNumber) => ({
    type: types.UPDATE_INPUT_VALUE,
    data: {postNumber}
});

export const getPosts = (posts) => ({
    type: types.GET_POSTS,
    data: posts
});

export const doGetPosts = (postId) => {
    return (dispatch, getState) => {
        return axios.get(`https://jsonplaceholder.typicode.com/posts/${postId}`)
            .then((response) => {
                dispatch(getPosts(response.data));
            })
            .catch((error) => {
                return new Error(error);
            });
    };
};