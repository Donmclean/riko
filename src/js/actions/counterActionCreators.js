import * as types from '../../js/constants/actions/actionTypes';

export const shouldIncrement = () => ({
    type: types.SHOULD_INCREMENT
});

export const increment = () => ({
    type: types.INCREMENT
});

export const shouldDecrement = () => ({
    type: types.SHOULD_DECREMENT
});

export const decrement = () => ({
    type: types.DECREMENT
});