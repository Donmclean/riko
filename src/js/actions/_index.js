import * as counterActionCreators from './counterActionCreators';
import * as asyncActionCreators from './asyncActionCreators';

const actionCreators = {
    counterActionCreators,
    asyncActionCreators
};

export default actionCreators;