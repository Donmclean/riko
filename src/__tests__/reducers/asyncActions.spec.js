"use strict";

import React from 'react';
import _ from 'lodash';
import * as asyncActions from '../../js/actions/asyncActionCreators';
import * as types from '../../js/constants/actions/actionTypes';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

// Useful for not making direct api requests when testing
// import nock from 'nock';

const middlewares = [ thunk ];
const mockStore = configureMockStore(middlewares);

describe('Async Actions', () => {
    it('setPosts should create a get posts action', () => {
        const expectedAction = {
            type: types.UPDATE_POSTS
        };

        expect(asyncActions.setPosts()).toEqual(expectedAction);
    });

    it('fetchPosts should return a function', () => {
        const postId = 1;
        const dispatch = asyncActions.fetchPosts(postId);
        expect(_.isFunction(dispatch)).toEqual(true);
    });

    it('fetchPosts creates UPDATE_POSTS when fetching posts has been done', () => {
        const postId = 1;
        const expectedData = [{
            type: 'UPDATE_POSTS',
            data: {
                userId: 1,
                id: 1,
                title: 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit',
                body: 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'
            }
        }];

        const store = mockStore();

        return store.dispatch(asyncActions.fetchPosts(postId))
            .then(() => { // return of async actions
                expect(store.getActions()).toEqual(expectedData);
            });
    });

    it('fetchPosts should return empty array upon error occurring when fetching invalid post number or (404)', () => {
        const postId = 0;
        const store = mockStore();

        return store.dispatch(asyncActions.fetchPosts(postId))
            .then(() => { // return of async actions
                expect(store.getActions()).toEqual([]);
            });
    });

    it('updatePostNumber should return valid data object', () => {
        const postNumber = 10;

        const dispatch = asyncActions.updatePostNumber(postNumber);
        const expectedDispatch = { type: 'UPDATE_POST_NUMBER', data: 10 };

        expect(dispatch).toEqual(expectedDispatch);
    });
});