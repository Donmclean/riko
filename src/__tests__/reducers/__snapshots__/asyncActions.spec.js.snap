exports[`Async Test Suite Async Dispatch Actions fetchPosts creates UPDATE_POSTS when fetching posts has been done 1`] = `
Array [
  Object {
    "data": Object {
      "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
      "id": 1,
      "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
      "userId": 1,
    },
    "type": "UPDATE_POSTS",
  },
]
`;
